### PR TITLE
Allow fragments to be imported by name when using the webpack loader

### PR DESCRIPTION
**Pull request by dobesv**
This makes fragments a bit more convenient to use.

Any packages that use `graphql-tag/loader` under the hood will also benefit.

Likely fix for https://github.com/apollographql/graphql-tag/issues/102 and https://github.com/apollographql/graphql-tag/issues/251

Note that if the fragment references other fragments (using fragment spread), you will get a document that has multiple fragments in it.  In this case you will have to tell apollo which fragment you want using the `fragmentName` option.

### CODE REVIEW

**Comment by gabsprates** [comment]
Looks good.
I had a problem here when tried to use the `readFragment()` method using a fragment imported and processed by loader.
Any updates?

**Comment by emahuni** [comment]
why isn't this being merged?

**Comment by silesky** [comment]
Can we merge this, please?

**Comment by dobesv** [comment]
@gabsprates I wasn't sure how to interpret your comment - are you saying there is an issue with this PR or that this PR would have solved an issue you had?

**Comment by denkristoffer** [comment]
@jnwng: Friendly ping, is anything missing here? 🙂

**Comment by dulakm** [comment]
Please, what's the status on this PR?

**Comment by ValdoGhafoor** [comment]
@jnwng A new year, a new ping for this PR

**Comment by fsjsd** [comment]
ping, need this

**Comment by jnwng** [comment]
@dobesv if you’re able, please add a couple of tests verifying the behavior you’ve added here, as well as an addition to the docs to let people know how this works. thank you in advance!

**Comment by dobesv** [comment]
Are there currently any tests for `graphql-tag/loader` ?  I didn't find any from a quick look in the `test` folder.

**Comment by jnwng** [comment]
here’s an example of the loader being tested to expose the correct names: https://github.com/apollographql/graphql-tag/blob/master/test/graphql.js#L71

**Comment by dobesv** [comment]
I added tests and updated the README.  I can't quite tell why the tests are failing, the tests that fail don't seem related to my changes.

**Comment by dobesv** [comment]
Tests are passing when I run them on my dev machine.  Anyone else who if watching this thread who might know why tests are behaving differently on Travis?  Maybe a node version issue?

**Comment by juliovedovatto** [comment]
I would love to use such feature. My project relies a lot on fragments and importing those fragments would save us time (also more organized).

Some points:

* I cloned @dobesv repo and ran under node@13. All tests passed
* I created a Docker, to test under node@10 and all tests passed
* Also tested under Docker node@8 and all tests passed

I believe something went wrong with Travis CI @jnwng 

My Docker setup:

*Dockerfile*

```
FROM node:10.8.0

RUN git clone https://github.com/dobesv/graphql-tag.git /app

WORKDIR /app
```
*docker-compose.yml*
```
version: "3.7"
services:
  graphql-tag:
    build:
      context: ./
    command: bash -c "cd /app && git checkout fragment-named-exports && git pull origin fragment-named-exports && rm -rf node_modules && yarn && yarn test"
    environment:
      NODE_ENV: development
```

Output:

```
graphql-tag_1  | Switched to a new branch 'fragment-named-exports'
graphql-tag_1  | Branch fragment-named-exports set up to track remote branch fragment-named-exports from origin.
graphql-tag_1  | From https://github.com/dobesv/graphql-tag
graphql-tag_1  |  * branch            fragment-named-exports -> FETCH_HEAD
graphql-tag_1  | Already up-to-date.
graphql-tag_1  | yarn install v1.9.2
graphql-tag_1  | [1/4] Resolving packages...
graphql-tag_1  | [2/4] Fetching packages...
graphql-tag_1  | (node:40) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
graphql-tag_1  | [3/4] Linking dependencies...
graphql-tag_1  | [4/4] Building fresh packages...
graphql-tag_1  | $ npm run bundle
graphql-tag_1  |
graphql-tag_1  | > graphql-tag@2.10.1 bundle /app
graphql-tag_1  | > rollup -c && cp src/index.js.flow lib/graphql-tag.umd.js.flow
graphql-tag_1  |
graphql-tag_1  | Done in 7.34s.
graphql-tag_1  | yarn run v1.9.2
graphql-tag_1  | $ mocha test/graphql.js test/graphql-v0.12.js && tav --ci --compat
graphql-tag_1  |
graphql-tag_1  |
graphql-tag_1  |   gql 0
graphql-tag_1  |     ✓ parses queries
graphql-tag_1  |     ✓ parses queries when called as a function
graphql-tag_1  |     ✓ parses queries with weird substitutions
graphql-tag_1  |     ✓ allows interpolation of documents generated by the webpack loader
graphql-tag_1  |     ✓ parses queries through webpack loader
graphql-tag_1  |     ✓ parses single query through webpack loader
graphql-tag_1  |     ✓ parses single query and exports as default
graphql-tag_1  |     ✓ parses multiple queries through webpack loader
graphql-tag_1  |     ✓ parses fragments with variable definitions
graphql-tag_1  |     ✓ does not nest queries needlessly in named exports
graphql-tag_1  |     ✓ tracks fragment dependencies from multiple queries through webpack loader
graphql-tag_1  |     ✓ tracks fragment dependencies across nested fragments
graphql-tag_1  |     ✓ correctly imports other files through the webpack loader
graphql-tag_1  |     ✓ tracks fragment dependencies across fragments loaded via the webpack loader
graphql-tag_1  |     ✓ does not complain when presented with normal comments
graphql-tag_1  |     ✓ returns the same object for the same query
graphql-tag_1  |     ✓ returns the same object for the same query, even with whitespace differences
graphql-tag_1  |     ✓ returns the same object for the same fragment
graphql-tag_1  |     ✓ returns the same object for the same document with substitution
graphql-tag_1  |     ✓ can reference a fragment that references as fragment
graphql-tag_1  |     fragment warnings
graphql-tag_1  |       ✓ warns if you use the same fragment name for different fragments
graphql-tag_1  |       ✓ does not warn if you use the same fragment name for the same fragment
graphql-tag_1  |       ✓ does not warn if you use the same embedded fragment in two different queries
graphql-tag_1  |       ✓ does not warn if you use the same fragment name for embedded and non-embedded fragments
graphql-tag_1  |     unique fragments
graphql-tag_1  |       ✓ strips duplicate fragments from the document
graphql-tag_1  |       ✓ ignores duplicate fragments from second-level imports when using the webpack loader
graphql-tag_1  |
graphql-tag_1  |   gql 1
graphql-tag_1  |     ✓ parses queries
graphql-tag_1  |     ✓ parses queries when called as a function
graphql-tag_1  |     ✓ parses queries with weird substitutions
graphql-tag_1  |     ✓ allows interpolation of documents generated by the webpack loader
graphql-tag_1  |     ✓ parses queries through webpack loader
graphql-tag_1  |     ✓ parses single query through webpack loader
graphql-tag_1  |     ✓ parses single query and exports as default
graphql-tag_1  |     ✓ parses multiple queries through webpack loader
graphql-tag_1  |     ✓ parses fragments with variable definitions
graphql-tag_1  |     ✓ does not nest queries needlessly in named exports
graphql-tag_1  |     ✓ tracks fragment dependencies from multiple queries through webpack loader
graphql-tag_1  |     ✓ tracks fragment dependencies across nested fragments
graphql-tag_1  |     ✓ correctly imports other files through the webpack loader
graphql-tag_1  |     ✓ tracks fragment dependencies across fragments loaded via the webpack loader
graphql-tag_1  |     ✓ does not complain when presented with normal comments
graphql-tag_1  |     ✓ returns the same object for the same query
graphql-tag_1  |     ✓ returns the same object for the same query, even with whitespace differences
graphql-tag_1  |     ✓ returns the same object for the same fragment
graphql-tag_1  |     ✓ returns the same object for the same document with substitution
graphql-tag_1  |     ✓ can reference a fragment that references as fragment
graphql-tag_1  |     fragment warnings
graphql-tag_1  |       ✓ warns if you use the same fragment name for different fragments
graphql-tag_1  |       ✓ does not warn if you use the same fragment name for the same fragment
graphql-tag_1  |       ✓ does not warn if you use the same embedded fragment in two different queries
graphql-tag_1  |       ✓ does not warn if you use the same fragment name for embedded and non-embedded fragments
graphql-tag_1  |     unique fragments
graphql-tag_1  |       ✓ strips duplicate fragments from the document
graphql-tag_1  |       ✓ ignores duplicate fragments from second-level imports when using the webpack loader
graphql-tag_1  |
graphql-tag_1  |   gql 0
graphql-tag_1  |     ✓ is correct for a simple query
graphql-tag_1  |     ✓ is correct for a fragment
graphql-tag_1  |
graphql-tag_1  |   gql 1
graphql-tag_1  |     ✓ is correct for a simple query
graphql-tag_1  |     ✓ is correct for a fragment
graphql-tag_1  |
graphql-tag_1  |
graphql-tag_1  |   56 passing (97ms)
graphql-tag_1  |
graphql-tag_1  | Done in 1.83s.
graphql-tag_graphql-tag_1 exited with code 0
```

**Comment by abdavid** [comment]
Yet another friendly ping :)

**Comment by MGray55** [comment]
Is this just pending due to merge conflicts?

**Comment by dobesv** [comment]
I rebased on master, pushed, and tested again.  Looks like the tests are passing now, which is great.

@jnwng any chance you would consider merging this PR now?

**Comment by xjunior** [comment]
Any chance this is going in any soon? In my experience importing graphql queries and not being able to mix them with fragments makes the use of `graphql-tag/loader` very limited.

**Comment by dobesv** [comment]
Note that as a workaround you can import a query document that uses the fragment you want to use, and use the `fragmentName` option to `readFragment` or `writeFragment` to indicate the fragment.  This change does look better in the code, though.

**Comment by thiagopradi** [comment]
@jnwng - is there anything that I can do to help merge this PR? happy to help if needed.

**Comment by lucasconstantino** [comment]
For those who don't wanna fork and still desire to use this: you can patch `graphql-tag` with [patch-package](https://www.npmjs.com/package/patch-package). After you install this, create a file at `./patches/graphql-tag+2.11.0.patch` with this content:

```patch
diff --git a/node_modules/graphql-tag/README.md b/node_modules/graphql-tag/README.md
index 56eb23b..f69a219 100644
--- a/node_modules/graphql-tag/README.md
+++ b/node_modules/graphql-tag/README.md
@@ -193,6 +193,32 @@ For **create-react-app** < v2, you'll either need to eject or use [react-app-rew
 
 Testing environments that don't support Webpack require additional configuration. For [Jest](https://facebook.github.io/jest/) use [jest-transform-graphql](https://github.com/remind101/jest-transform-graphql).
 
+#### Support for fragments
+
+With the webpack loader, you can import fragments by name:
+
+In a file called `query.gql`:
+```graphql
+fragment MyFragment1 on MyType1 {
+  ...
+}
+
+fragment MyFragment2 on MyType2 {
+  ...
+}
+```
+
+And in your JavaScript:
+
+```javascript
+import { MyFragment1, MyFragment2 } from 'query.gql'
+```
+
+Note: If your fragment references other fragments, the resulting document will
+have multiple fragments in it.  In this case you must still specify the fragment
+name when using the fragment.  For example, with `apollo-client` you would specify
+the `fragmentName` option when using the fragment for cache operations. 
+
 ### Warnings
 
 This package will emit a warning if you have multiple fragments of the same name. You can disable this with:
diff --git a/node_modules/graphql-tag/loader.js b/node_modules/graphql-tag/loader.js
index 4ceac4d..5f554e5 100644
--- a/node_modules/graphql-tag/loader.js
+++ b/node_modules/graphql-tag/loader.js
@@ -53,7 +53,7 @@ module.exports = function(source) {
   // at compile time, and then uses those at load time to create minimal query documents
   // We cannot do the latter at compile time due to how the #import code works.
   let operationCount = doc.definitions.reduce(function(accum, op) {
-    if (op.kind === "OperationDefinition") {
+    if (op.kind === "OperationDefinition" || op.kind === "FragmentDefinition") {
       return accum + 1;
     }
 
@@ -166,7 +166,7 @@ module.exports = function(source) {
     `
 
     for (const op of doc.definitions) {
-      if (op.kind === "OperationDefinition") {
+      if (op.kind === "OperationDefinition" || op.kind === "FragmentDefinition") {
         if (!op.name) {
           if (operationCount > 1) {
             throw "Query/mutation names are required for a document with multiple definitions";
diff --git a/node_modules/graphql-tag/test/graphql.js b/node_modules/graphql-tag/test/graphql.js
index e77af23..1c4c8f8 100644
--- a/node_modules/graphql-tag/test/graphql.js
+++ b/node_modules/graphql-tag/test/graphql.js
@@ -141,7 +141,18 @@ describe('gql', () => {
     assert.equal(Q3[1].name.value, 'F1');
     assert.equal(Q3[2].name.value, 'F2');
 
-  });
+      const F1 = module.exports.F1.definitions;
+      const F2 = module.exports.F2.definitions;
+      const F3 = module.exports.F3.definitions;
+
+      assert.equal(F1.length, 1);
+      assert.equal(F1[0].name.value, 'F1');
+      assert.equal(F2.length, 1);
+      assert.equal(F2[0].name.value, 'F2');
+      assert.equal(F3.length, 1);
+      assert.equal(F3[0].name.value, 'F3');
+
+    });
 
   it('tracks fragment dependencies across nested fragments', () => {
     const jsSource = loader.call({ cacheable() {} }, `
@@ -179,8 +190,22 @@ describe('gql', () => {
     assert.equal(Q1[2].name.value, 'F22');
     assert.equal(Q1[3].name.value, 'F11');
 
-    assert.equal(Q2.length, 1);
-  });
+      assert.equal(Q2.length, 1);
+
+      const F11 = module.exports.F11.definitions;
+      const F22 = module.exports.F22.definitions;
+      const F33 = module.exports.F33.definitions;
+
+      assert.equal(F11.length, 1);
+      assert.equal(F11[0].name.value, 'F11');
+      assert.equal(F22.length, 2);
+      assert.equal(F22[0].name.value, 'F22');
+      assert.equal(F22[1].name.value, 'F11');
+      assert.equal(F33.length, 3);
+      assert.equal(F33[0].name.value, 'F33');
+      assert.equal(F33[1].name.value, 'F22');
+      assert.equal(F33[2].name.value, 'F11');
+    });
 
   it('correctly imports other files through the webpack loader', () => {
     const query = `#import "./fragment_definition.graphql"

```

**Comment by webdeb** [comment]
#331 for ref

**Comment by dobesv** [comment]
If you're using yarn 2 it has a built in package patching method you can use instead of patch-package

File: loader.js
Comment (webdeb) [inline loader.js:164]: I can imagine, that this line produces a mental dissonance on the apollo team. They don't know exactly why they can't merge this, they just know that they cannot merge it. 

@dobesv removing this line will make wonders

**Comment by uproot-co** [comment]
Are there any plans to merge this, please?

**Comment by flzozaya** [comment]
Are there any plans to merge this, please?

**Comment by Vincz** [comment]
Hi guys. This feature is really needed with the new cache modification system (readFragment, writeFragment, ...).
And it's just two lines of code ...

ping @jnwng 
ping @benjamn 
ping @hwillson

**Review by hwillson — APPROVED** [review]
Thanks very much @dobesv!

**Comment by hwillson** [comment]
Available in `graphql-tag@2.12.4`.

**Comment by Vincz** [comment]
Thank you @hwillson !

**Pull request merged**
